File: C#/LocalSearchOptimization/Components/Operator.cs
﻿using System.Collections.Generic;

namespace LocalSearchOptimization.Components
{
    public abstract class Operator
    {
        public double Weight { get; set; }

        public List<Configuration> Configurations { get; }

        public int Power
        {
            get
            {
                return Configurations.Count;
            }
        }

        public Operator(double weight)
        {
            Weight = weight;
            Configurations = new List<Configuration>();
        }

        public abstract ISolution Apply(ISolution solution, Configuration configuration);

        public ISolution Apply(ISolution solution, int configuration)
        {
            return Apply(solution, Configurations[configuration]);
        }
    }
}

File: java/src/localsearchoptimization/components/Operator.java
package localsearchoptimization.components;

import java.util.ArrayList;
import java.util.List;

public abstract class Operator {

    public double weight;

    public List<Configuration> configurations;

    public int power() {
        return configurations.size();
    }

    public Operator(double weight) {
        this.weight = weight;
        configurations = new ArrayList<Configuration>();
    }

    public abstract Solution apply(Solution solution, Configuration configuration);

    public Solution apply(Solution solution, int configuration) {
        return this.apply(solution, configurations.get(configuration));
    }
}